packages/server/src/time-control.ts

```diff
@@ -40,100 +54,326 @@ export function ValidateTimeControlBase(
 export interface ITimeHandler {
+  /**
+   * Returns the initial state of a time control data object
+   * that this handler works with
+   */
+  initialState(
+    variant: string,
+    config: IConfigWithTimeControl,
+  ): ITimeControlBase;
+
+  /**
+   * transition for the time control data when a move is played
+   */
   handleMove(
     game: GameResponse,
     game_obj: AbstractGame<unknown, unknown>,
     playerNr: number,
     move: string,
-  ): Promise<void>;
+  ): ITimeControlBase;
+
+  /**
+   * Returns the time in milliseconds until this player
+   * times out, provided no moves are played
+   */
+  getMsUntilTimeout(game: GameResponse, playerNr: number): number;
 }
 
 class TimeHandlerSequentialMoves implements ITimeHandler {
-  async handleMove(
-    game: GameResponse,
-    game_obj: AbstractGame<unknown, unknown>,
-    playerNr: number,
-  ): Promise<void> {
-    if (!HasTimeControlConfig(game.config)) {
-      console.log("has no time control config");
+  private _timeoutService: TimeoutService;
+  constructor() {
+    this._timeoutService = getTimeoutService();
+  }
 
-      return;
+  initialState(
+    variant: string,
+    config: IConfigWithTimeControl,
+  ): ITimeControlBase {
+    const numPlayers = makeGameObject(variant, config).numPlayers();
+
+    const timeControl: ITimeControlBase = {
+      moveTimestamps: [],
+      forPlayer: {},
+    };
+
+    for (let i = 0; i < numPlayers; i++) {
+      timeControl.forPlayer[i] = {
+        remainingTimeMS: config.time_control.mainTimeMS,
+        onThePlaySince: null,
+      };
     }
 
-    switch (game.config.time_control.type) {
+    switch (config.time_control.type) {
       case TimeControlType.Absolute: {
-        let timeData: ITimeControlBase = game.time_control;
-        if (timeData === undefined) {
-          timeData = {
-            moveTimestamps: [],
-            forPlayer: {},
-          };
-        }
+        // nothing to do
+        break;
+      }
 
-        if (timeData.forPlayer[playerNr] === undefined) {
-          timeData.forPlayer[playerNr] = {
-            remainingTimeMS: game.config.time_control.mainTimeMS,
-            onThePlaySince: undefined,
-          };
-        }
+      case TimeControlType.Fischer: {
+        // nothing to do
+        break;
+      }
 
-        const nextPlayers = game_obj.nextToPlay();
+      default: {
+        console.error("received config with invalid time control type");
+      }
+    }
 
-        if (timeData.forPlayer[playerNr] === null) {
-          console.error(`game with id ${game.id} has defect time control data`);
-          return;
-        }
+    return timeControl;
+  }
 
-        const playerData = timeData.forPlayer[playerNr];
+  handleMove(
+    game: GameResponse,
+    game_obj: AbstractGame<unknown, unknown>,
+    playerNr: number,
+  ): ITimeControlBase {
+    const config = game.config as IConfigWithTimeControl;
 
+    switch (config.time_control.type) {
```

Comment:
I removed it in the handleMove method, because the playMove function in games.ts already validates the time control config before this. Maybe it could also be removed in the getMsUntilTimeout method, but that is only being called on startup at the moment.

Reply:
Alright, thanks